### PR TITLE
feat(gatsby-remark-images): Add flag to suppress css background-image to prevent FOUB

### LINKED ISSUE #16516: gatsby-remark-images: disable background-image on gatsby-resp-image-background-image to prevent FOUB
## Description

[gatsby-remark-images](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-remark-images) has a blur-up effect that does not work well on `.png` images with transparent edges. I looked at the gatsby-remark-images [source code](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-remark-images/src/gatsby-browser.js#L23) and it is using css transition for the blur up effect. Unfortunately, the css transitions does not work well on images with alpha transparent edges. It causes a flash of ugly blur-up (FOUB) effect. We need a way to prevent/disable `background-image` css property on `gatsby-resp-image-background-image` class for `.png`s with transparent edges to prevent FOUB.

Related to https://github.com/gatsbyjs/gatsby/issues/4510#issuecomment-390781917

(**Note:** the blur-up effect doesn't work well even with `backgroundColor: 'transparent'` option set for `gatsby-remark-images`)

### Screenshots

white-no-box-shadow-error
black-no-box-shadow-error
png-transparency-blur-issue

### gatsby-config

```js
{
  resolve: `gatsby-transformer-remark`,
  options: {
    plugins: [
      {
        resolve: `gatsby-remark-images`,
        options: {
          maxWidth: 590,
          backgroundColor: `transparent`,
        },
      },
    ],
  },
},
```

We should have the ability to disable background-image css property on `gatsby-resp-image-background-image` class on images if it contains alpha transparency.

I think we should automatically disable background-image css property for `.png` images if  `backgroundColor: 'transparent"` option is set. We can do this for all .png images or only for .png images that have alpha transparency. 

### Steps to reproduce

- Create a gatsby project.
- Install and add `gatsby-transformer-remark` and `gatsby-remark-images` plugins
- set `backgroundColor: 'transparent'` option for `gatsby-remark-images` plugin
- add .png images to your markdown files that contain alpha transparent edges
- open webpage and refresh page to see flash of ugly blur-up (FOUB)

### Expected result

There is no flash of ugly blur-up (FOUB) on page refresh.

### Actual result

There is flash of ugly blur-up (FOUB) on page refresh.

### Link to demo project

- Go to https://paulkode.com/git-cannot-squash-without-a-previous-commit/
- Refresh the page (try in both dark/light modes)

### Environment

```
  System:
    OS: Linux 4.15 Ubuntu 18.04.3 LTS (Bionic Beaver)
    CPU: (2) x64 Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz
    Shell: 4.4.20 - /bin/bash
  Binaries:
    Node: 12.2.0 - ~/.nvm/versions/node/v12.2.0/bin/node
    Yarn: 1.17.3 - ~/.nvm/versions/node/v12.2.0/bin/yarn
    npm: 6.10.3 - ~/.nvm/versions/node/v12.2.0/bin/npm
  Languages:
    Python: 3.7.3 - /home/paul/.pyenv/shims/python
  Browsers:
    Chrome: 76.0.3809.100
    Firefox: 68.0.1
  npmPackages:
    gatsby: ^2.13.57 => 2.13.57
    gatsby-image: ^2.2.8 => 2.2.8
    gatsby-plugin-emotion: ^4.1.2 => 4.1.2
    gatsby-plugin-feed: ^2.3.6 => 2.3.6
    gatsby-plugin-google-analytics: ^2.1.7 => 2.1.7
    gatsby-plugin-manifest: ^2.2.5 => 2.2.5
    gatsby-plugin-offline: ^2.2.5 => 2.2.5
    gatsby-plugin-react-helmet: ^3.1.3 => 3.1.3
    gatsby-plugin-sass: ^2.1.6 => 2.1.6
    gatsby-plugin-sharp: ^2.2.11 => 2.2.11
    gatsby-plugin-styled-components: ^3.1.2 => 3.1.2
    gatsby-remark-copy-linked-files: ^2.1.5 => 2.1.5
    gatsby-remark-external-links: 0.0.4 => 0.0.4
    gatsby-remark-images: ^3.1.8 => 3.1.8
    gatsby-remark-katex: ^3.1.2 => 3.1.2
    gatsby-remark-prismjs: ^3.3.5 => 3.3.5
    gatsby-remark-responsive-iframe: ^2.2.4 => 2.2.4
    gatsby-remark-smartypants: ^2.1.2 => 2.1.2
    gatsby-source-filesystem: ^2.1.9 => 2.1.9
    gatsby-transformer-remark: ^2.6.12 => 2.6.12
    gatsby-transformer-sharp: ^2.2.6 => 2.2.6
```

### PR DESCRIPTION
**Pull request by kimbaudi**
## Description

This PR makes changes to [gatsby-plugin-sharp](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-plugin-sharp) and [gatsby-remark-images](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-remark-images) in order to address issue #16516.

I created a new option for `gatsby-remark-images` plugin called `disableBgImageOnAlpha` which is disabled by default. When `disableBgImageOnAlpha` is enabled (i.e., set to `true`), it will check the image file for opacity/transparency. If the image contains transparency, then the `background-image`  css property is disabled on `<span class="gatsby-resp-image-background-image"></span>`, which will prevent the "flash of ugly blur-up" (FOUB). 

The reason why I made changes to `gatsby-plugin-sharp` is to expose [stats](http://sharp.pixelplumbing.com/en/stable/api-input/#stats), which is how I determine whether the image contains transparency.

## Related Issues

Fixes #16516

### CODE REVIEW

File: packages/gatsby-remark-images/src/index.js [hunk 1]
```diff
@@ -282,10 +282,20 @@ module.exports = (
...
+
+    const bgImage = !removeBgImage
+      ? ` background-image: url('${placeholderImageData}');`
+      : ``
+
     let rawHTML = `
   <span
     class="${imageBackgroundClass}"
-    style="padding-bottom: ${ratio}; position: relative; bottom: 0; left: 0; background-image: url('${placeholderImageData}'); background-size: cover; display: block;"
+    style="padding-bottom: ${ratio}; position: relative; bottom: 0; left: 0;${bgImage} background-size: cover; display: block;"
```
Comment (wardpeet) [inline packages/gatsby-remark-images/src/index.js]: we can move `background-size: cover;` also inside the bgImage variable as it won't have any effect when there is no background-image set.

File: packages/gatsby-remark-images/src/index.js [hunk 2]
```diff
@@ -282,10 +282,20 @@ module.exports = (
     const imageCaption =
       options.showCaptions && getImageCaption(node, overWrites)
 
+    let removeBgImage = false
+    if (options.disableBgImageOnAlpha) {
+      const imageStats = await stats({ file: imageNode, reporter })
+      if (imageStats && imageStats.isTransparent) removeBgImage = true
+    }
+
+    const bgImage = !removeBgImage
```
[the remark this answers is not in the document]
Comment (kimbaudi) [reply packages/gatsby-remark-images/src/index.js]: i moved the truthy condition first

File: packages/gatsby-remark-images/src/index.js (hunk 1 above)
Comment (kimbaudi) [reply packages/gatsby-remark-images/src/index.js]: i moved `background-size: cover;` inside the `bgImage` variable

**Comment by kimbaudi** [comment]
Thanks for the code review. I've made the necessary changes.

**Comment by kimbaudi** [comment]
Currently, there is 1 failing test in `packages/gatsby/src/bootstrap/load-plugins/__tests__/load-plugins.js`. I hope it doesn't affect this PR getting merged 😄.

Here 

```
Summary of all failing tests
FAIL packages/gatsby/src/bootstrap/load-plugins/__tests__/load-plugins.js (5.042s)
  ● Load plugins › Overrides the options for gatsby-plugin-page-creator

    expect(received).toMatchSnapshot()

    Snapshot name: `Load plugins Overrides the options for gatsby-plugin-page-creator 1`

    - Snapshot
    + Received

    @@ -89,22 +89,10 @@
          "version": "1.0.0",
        },
        Object {
          "browserAPIs": Array [],
          "id": "",
    -     "name": "default-site-plugin",
    -     "nodeAPIs": Array [],
    -     "pluginOptions": Object {
    -       "plugins": Array [],
    -     },
    -     "resolve": "",
    -     "ssrAPIs": Array [],
    -     "version": "1.0.0",
    -   },
    -   Object {
    -     "browserAPIs": Array [],
    -     "id": "",
          "name": "gatsby-plugin-page-creator",
          "nodeAPIs": Array [
            "createPagesStatefully",
          ],
          "pluginOptions": Object {
    @@ -197,17 +185,13 @@
          "version": "1.0.0",
        },
        Object {
          "browserAPIs": Array [],
          "id": "",
    -     "name": "gatsby-plugin-page-creator",
    -     "nodeAPIs": Array [
    -       "createPagesStatefully",
    -     ],
    +     "name": "default-site-plugin",
    +     "nodeAPIs": Array [],
          "pluginOptions": Object {
    -       "path": "<PROJECT_ROOT>/src/pages",
    -       "pathCheck": false,
            "plugins": Array [],
          },
          "resolve": "",
          "ssrAPIs": Array [],
          "version": "1.0.0",

      70 |     plugins = replaceFieldsThatCanVary(plugins)
      71 | 
    > 72 |     expect(plugins).toMatchSnapshot()
         |                     ^
      73 |   })
      74 | })
      75 | 

      at Object.toMatchSnapshot (packages/gatsby/src/bootstrap/load-plugins/__tests__/load-plugins.js:72:21)
```

**Comment by wardpeet** [comment]
ok sweet, thanks. Doing a quick test and I'll merge right after! Thank you for doing this and for your patience.

**Comment by kimbaudi** [comment]
@wardpeet - Although this PR got merged into master, and [gatsby-remark-images doc](https://www.gatsbyjs.org/packages/gatsby-remark-images/) shows the new `disableBgImageOnAlpha` option, this PR is not available. Seems like some checks haven't completed (https://github.com/gatsbyjs/gatsby/commit/125fd01ee186577c912a33c3950b345395cda76a). I just want to know if you are aware of this. Hopefully, all check will be completed soon :smile:

**Comment by kimbaudi** [comment]
I just confirmed that this PR got merged into master 😄

**Pull request merged**